### ISSUE TITLE
TestCudaLocalEnergyMinimizer mixed fails

### ISSUE THREAD

**Issue by swails** [pandegroup/openmm#867]
```
swails@batman ~/src/build_openmm_dev $ ./TestCudaLocalEnergyMinimizer single
Done
swails@batman ~/src/build_openmm_dev $ ./TestCudaLocalEnergyMinimizer mixed
exception: Assertion failure at TestCudaLocalEnergyMinimizer.cpp:133
swails@batman ~/src/build_openmm_dev $ ./TestCudaLocalEnergyMinimizer double
Done
```

This is reproducible -- mixed precision always assert-fails at the same line.  If @peastman (or whoever goes about debugging this) can't repro, I'll provide more info.

**Comment by swails** [comment]
This is still happening for me with the latest code.

**Comment by swails** [comment]
There is definitely an issue here -- the `forceNorm` (net normalized force?) in the mixed precision is over 3x larger than that in single precision! (~5.1 for single, >17 for mixed).  Double is, not surprisingly, even lower (ca. 1)

**Comment by peastman** [comment]
I'll try to look into this soon.

**Comment by swails** [comment]
I wasn't trying to be pushy (sorry if I came off that way).  The above was just a mind dump and (some) details of what I did to look into the problem -- this is not a pressing need for me (although I'm curious about the culprit).

**Comment by peastman** [comment]
No problem!  I'm just a little bit buried in a stack of other things, or I would have investigated it already.

**Comment by swails** [comment]
I figured as much

**Comment by peastman** [comment]
I can't reproduce this.  I just tried on two different computers, and it passed repeatedly on both of them.  Does anyone else see this?

**Comment by kyleabeauchamp** [comment]
```
kyleb@jak:~/src/openmm/omm-build$ ./TestCudaLocalEnergyMinimizer single
Done
kyleb@jak:~/src/openmm/omm-build$ ./TestCudaLocalEnergyMinimizer mixed
exception: Assertion failure at TestCudaLocalEnergyMinimizer.cpp:133
kyleb@jak:~/src/openmm/omm-build$ ./TestCudaLocalEnergyMinimizer double
Done
kyleb@jak:~/src/openmm/omm-build$ nvidia-smi 
Mon Apr 27 20:19:16 2015       
+------------------------------------------------------+                       
| NVIDIA-SMI 340.29     Driver Version: 340.29         |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TITAN   Off  | 0000:03:00.0     N/A |                  N/A |
| 42%   68C    P0    N/A /  N/A |    338MiB /  6143MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX TITAN   Off  | 0000:83:00.0     N/A |                  N/A |
| 40%   65C    P0    N/A /  N/A |     90MiB /  6143MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce GTX TITAN   Off  | 0000:84:00.0     N/A |                  N/A |
| 48%   75C    P0    N/A /  N/A |     89MiB /  6143MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Compute processes:                                               GPU Memory |
|  GPU       PID  Process name                                     Usage      |
|=============================================================================|
|    0            Not Supported                                               |
|    1            Not Supported                                               |
|    2            Not Supported                                               |
+-----------------------------------------------------------------------------+

```

**Comment by peastman** [comment]
I have a newer driver (346.46), but otherwise that looks very similar.  I tested on both a GTX Titan and a GTX 980.

**Comment by jchodera** [comment]
@kyleabeauchamp: `src` has the newer driver but CUDA 6.5, and a Titan and 980.  Can you try that when you get a chance?

**Comment by kyleabeauchamp** [comment]
```
kyleb@src:~/src/openmm/omm-build$ nvidia-smi 
Mon Apr 27 20:16:42 2015       
+------------------------------------------------------+                       
| NVIDIA-SMI 346.16     Driver Version: 346.16         |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX TITAN   Off  | 0000:04:00.0     Off |                  N/A |
| 64%   83C    P0   158W / 250W |    131MiB /  6143MiB |     97%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 980     Off  | 0000:41:00.0     N/A |                  N/A |
| 44%   78C    P2    N/A /  N/A |    139MiB /  4095MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0     25395    C   .../NVIDIA/Fermi/beta/Core_18.fah/FahCore_18   114MiB |
|    1            C+G   Not Supported                                         |
+-----------------------------------------------------------------------------+
kyleb@src:~/src/openmm/omm-build$ export CUDA_VISIBLE_DEVICES="0"
kyleb@src:~/src/openmm/omm-build$ ./TestCudaLocalEnergyMinimizer single
Done
kyleb@src:~/src/openmm/omm-build$ ./TestCudaLocalEnergyMinimizer mixed
exception: Assertion failure at TestCudaLocalEnergyMinimizer.cpp:133
kyleb@src:~/src/openmm/omm-build$ export CUDA_VISIBLE_DEVICES="1"
kyleb@src:~/src/openmm/omm-build$ ./TestCudaLocalEnergyMinimizer mixed
exception: Assertion failure at TestCudaLocalEnergyMinimizer.cpp:133
```

**Comment by swails** [comment]
Unfortunately the Rutgers network is currently experiencing issues, so I can't log in to my computer to check.

I'm running a GTX 680 with the 340.xx series drivers (I'm still using a GTS 250 to power my display, which is not supported by the 346 series :)).  I'm pretty sure that I tried checking out 6.2 and that passed.  So when I'm at my computer tomorrow I'll report the results of a bisect to nail down the offending commit.

**Comment by peastman** [comment]
One thing about this test case is that it's deterministic on any given computer (since it always initializes the random number generator to 0), but may produce different results on different computers.  I modified line 94 to pick a different random seed each time:

    init_gen_rand(time(NULL), sfmt);

(You'll also need to #include \<time.h\>.)  And then at the end of testLargeSystem() I print out forceNorm.  I ran it five times in each precision mode.  Here are the values:

Single:
4.43387
4.82453
7.18042
3.03302
6.3804

Mixed:
5.41214
5.16115
9.9919
5.74225
5.74469

Double:
2.46975
3.12669
2.28104
3.30251
2.39771

There doesn't seem to be any significant difference between single and mixed precision.  The values for double are a bit smaller.  It never came close to failing in any mode.  For that to happen, forceNorm would have to be 15.

What sorts of values are you getting?

**Comment by swails** [comment]
I *was* getting 17<forceNorm<18.  I'll try with your new code.

But where are random numbers used here??

**Comment by swails** [comment]
OK, setting random seeds, I get values 3<forceNorm<11.  In 30 trials, I do get 2 failures on this line:

```
    ASSERT(finalState.getPotentialEnergy() < initialState.getPotentialEnergy());
```

```
exception: Assertion failure at TestCudaLocalEnergyMinimizer.cpp:115
9.21616
3.62068
Done
exception: Assertion failure at TestCudaLocalEnergyMinimizer.cpp:115
```

**Comment by peastman** [comment]
I think we may simply be specifying a value for `tolerance` that is smaller than it can reliably achieve without using double precision.  I tried increasing it to 10, and found `forceNorm` actually got slightly smaller!  Presumably the minimizer was struggling to reach its goal, and making things worse in the process.  Try doing that, and then also modify the final assertion to

    ASSERT(forceNorm < 2*tolerance);

(It really should be strictly less than `tolerance`, but I'm willing to give it a little slack if it occassionally still has trouble reaching that.)  Does that now work reliably for you in all modes?

**Comment by jchodera** [comment]
> I think we may simply be specifying a value for tolerance that is smaller than it can reliably achieve without using double precision. 

Any intuition into why this might be?  It's not surprising that there is a difference between single and double precision, but it *is* surprising if the single precision gradient isn't sufficiently free from numerical noise to reliably point downhill...

**Comment by peastman** [comment]
I can think of a couple of reasons.  First, the minimizer always works in double precision.  That's how the library it uses works.  But force calculations are done in single precision.  So within a certain range, the minimizer will find that changing the coordinates has no effect on the force (because it's making tiny changes that get lost when converting to single).  That could certainly confuse it.

Second, we're dealing with an inherently difficult problem: getting large forces to perfectly cancel each other out.  That will necessarily be more difficult in single than in double precision, and beyond a certain point may just not be possible.

**Comment by jchodera** [comment]
> So within a certain range, the minimizer will find that changing the coordinates has no effect on the force (because it's making tiny changes that get lost when converting to single). That could certainly confuse it.

Does this suggest we should increase the `xtol` you set here to 1e-7 to eliminate this aspect of the problem?
https://github.com/pandegroup/openmm/blob/master/openmmapi/src/LocalEnergyMinimizer.cpp#L117-L118
```
    if (!context.getPlatform().supportsDoublePrecision())
        param.xtol = 1e-7;
```

**Comment by peastman** [comment]
Yes, changing that might well help.  I should really deprecate supportsDoublePrecision(), since it has very little meaning anymore.  We need some mechanism the minimizer could use to discover the appropriate precision to assume for a given context.  Currently there isn't one.

**Comment by swails** [comment]
> Yes, changing that might well help. I should really deprecate supportsDoublePrecision(), since it has very little meaning anymore. We need some mechanism the minimizer could use to discover the appropriate precision to assume for a given context. Currently there isn't one.

+1

I won't be able to check anything until tomorrow.  Rutgers' network is under a DDoS attack right now which makes it very hard to connect to my work computer.  But I have no reason to suspect that @peastman is wrong in his suspicions.

I'll try what @peastman suggested and then modify my PR accordingly (unless you want to issue a new PR on your own)

**Comment by peastman** [comment]
First let me see if I can come up with a better fix that actually improves the accuracy of the result.

**Comment by swails** [comment]
OK, I'm more than happy to leave it to you :).

**Comment by peastman** [comment]
Changing `xtol` doesn't seem to help.  It still just can't get that high accuracy in single or mixed precision.  So I guess I'll just change the test case.

This makes me wonder whether I should change the default value for `tolerance`.  Currently it's 1.  If that's unrealistic for typical cases, perhaps I should change it to 10?

**Comment by peastman** [comment]
I think there might also be a bug in the test case.  I'm looking at the line

    forceNorm = sqrt(forceNorm/(4*numMolecules));

These are diatomic molecules with a single constrained distance.  So shouldn't that be 5, not 4?  It only makes a small difference (forceNorm comes out 12% too high), and doesn't fix the problem.  But it does contribute a little to it.

**Comment by jchodera** [comment]
Gromacs also permits [L-BFGS optimization with single precision](http://www.gromacs.org/Documentation/Terminology/Energy_Minimisation).  I wonder if they have this problem too...

**Comment by peastman** [comment]
Fixed by #903.
